Bas tracing

bjoernschroeder opened the pull request.

bjoernschroeder committed:
implement planPool

bjoernschroeder committed:
start implementation of tracing

bjoernschroeder committed:
implement IAlicaTrace and IAlicaTraceFactory, add tracing for behaviours init/terminate

veerajsk commented on alica_engine/CMakeLists.txt:
```diff
@@ -10,16 +10,16 @@ find_package(yaml-cpp REQUIRED)
 find_package(catkin QUIET)
 if (catkin_FOUND)
-  find_package(catkin REQUIRED COMPONENTS id_manager fsystem event_handling system_util alica_common_config alica_solver_interface)
+  find_package(catkin REQUIRED COMPONENTS id_manager fsystem event_handling system_util alica_common_config alica_solver_interface amr_interfaces)
```
alica should not depend on amr_interfaces. I assume this is needed for the trace context. In that case, we simply use std::string directly.

veerajsk commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -73,6 +75,8 @@ class BasicBehaviour
 
     bool isEventDriven() const { return _behaviour->isEventDriven(); }
 
+    amr_interfaces::TraceContext getTraceContext() { return _trace->context(); };
```
can instead return a reference to the trace object itself i.e. IAlicaTrace&. That way the application can set logs & tags into the trace while the behaviour/plan is active.

veerajsk commented on alica_engine/include/engine/IAlicaTrace.h:
```diff
@@ -0,0 +1,32 @@
+{
+public:
+    virtual ~IAlicaTrace() = default;
+    IAlicaTrace(const std::string& opName, std::optional<const amr_interfaces::TraceContext> parent = std::nullopt);
```
the interface should not have any constructors, because the factory creates the traces. Use the same design as alica timer factory & alica timers

veerajsk commented on alica_engine/include/engine/IAlicaTrace.h:
```diff
@@ -0,0 +1,32 @@
+    IAlicaTrace(IAlicaTrace&& other) = default;
+    IAlicaTrace& operator=(IAlicaTrace&& other) = default;
+
+//    void setTag(const std::string& key, const RawTraceValue& value);
```
the interface can have 4 pure virtual functions:
```
virtual void setTag(const std::string& key, const std::string& value) = 0;
virtual void setLog(const std::string& key, const std::string& value) = 0;
virtual void markError(const std::string& description) = 0;
std::string context() const = 0;
```
Ideally setTag & setLog should be able to take ints, doubles etc. as values, but can figure that out later. string should be fine for now

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -185,4 +188,19 @@ bool BasicBehaviour::getParameter(const std::string& key, std::string& valueOut)
+{
+    RunningPlan* parent = _context.load()->getParent();
+    if (parent == nullptr) {
+        // a behaviour always has a plan as its parent
```
this should be an assert then

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -185,4 +188,19 @@ bool BasicBehaviour::getParameter(const std::string& key, std::string& valueOut)
+    }
+
+    if (auto parentPlan = parent->getBasicPlan()) {
+        _trace = std::make_unique<IAlicaTrace>("Behaviour", parentPlan->getTraceContext());
```
use the trace factory to create the trace. The name of the trace should be the behaviour name

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -112,6 +113,7 @@ void BasicBehaviour::initJob()
 
     // Todo: Optimization: don't call initialiseParameters if not in context
     try {
+        auto trace = IAlicaTrace("Init", _trace->context());
```
there is no need to create traces for init/terminate. The trace should be created in start() & destroyed in stop(). For init, terminate & the first run() call set corresponding boolean tags on the trace i.e. `_trace->setTag("init", "true")`.

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -185,4 +188,19 @@ bool BasicBehaviour::getParameter(const std::string& key, std::string& valueOut)
+
+    if (auto parentPlan = parent->getBasicPlan()) {
+        _trace = std::make_unique<IAlicaTrace>("Behaviour", parentPlan->getTraceContext());
+    } else {
```
remove this & assert that there is a parentPlan instead

veerajsk commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -23,6 +23,7 @@ void BasicPlan::doInit()
 {
     ++_execState;
     try {
+        auto trace = IAlicaTrace("Init", _trace->context());
```
same comments as for basic behaviour apply here as well

veerajsk commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -162,5 +167,6 @@ class BasicBehaviour
 
     int64_t _activeRunJobId;
     std::atomic<bool> _triggeredJobRunning;
+    std::unique_ptr<IAlicaTrace> _trace;
```
can use optional instead. That way we save on 1 dynamic memory allocation

veerajsk reviewed (changes requested):
Couple of general comments:
1. I'm not sure if you already have implementations for IAlicaTrace & IAlicaTraceFactory in supplementary. But we need them & you can use the same design as IAlicaTimer & IAlicaTimerFactory. Adapt amr tracing for the implementation such that it fits the interface required by IAlicaTrace.
2. I'm not sure why there are 359 files changed. Can we just have the tracing implementation changes in this PR, otherwise it becomes very difficult to review or merge.

bjoernschroeder committed:
replace tracecontext with string, change methods to pure virtual

bjoernschroeder committed:
add getter for traceFactory

bjoernschroeder committed:
replace getter for traceContext with getter for trace

bjoernschroeder committed:
use setLog for init/terminate traces, use factory to create traces

bjoernschroeder committed:
remove amr_interfaces dependency

bjoernschroeder committed:
add default constructor, make create pure virtual

bjoernschroeder committed:
return ptr to traceFactory

bjoernschroeder committed:
return ptr to traceFactory

bjoernschroeder committed:
add trace tag for run methods, check if trace / tracefactory exists before usage

bjoernschroeder committed:
add name to basicPlan, use name for tag

veerajsk commented on alica_engine/include/engine/IAlicaTrace.h:
```diff
@@ -0,0 +1,34 @@
+{
+public:
+    virtual ~IAlicaTrace() = default;
+    IAlicaTrace() = default;
```
there is no need for constructors in an interface class. The factory implementation is responsible for creating trace objects

veerajsk commented on alica_engine/include/engine/IAlicaTrace.h:
```diff
@@ -0,0 +1,34 @@
+    virtual std::string context() const = 0;
+
+private:
+    std::unique_ptr<IAlicaTrace> _rawTrace;
```
an interface class should not have any member variables. The implementation will do that. Also this is a recursive reference to another IAlicaTrace which I don't quite understand.

veerajsk commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -51,6 +57,9 @@ void BasicPlan::doTerminate()
     ++_execState;
     try {
+        if (_trace) {
+            _trace->setTag("Terminate", "true");
```
need to destroy the trace here

abhi1287 replied:
log instead of tag

bjoernschroeder committed:
remove member variable and constructors

bjoernschroeder committed:
destroy traces after setting terminate tag

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -73,6 +74,8 @@ class BasicBehaviour
 
     bool isEventDriven() const { return _behaviour->isEventDriven(); }
 
+    IAlicaTrace& getTrace() { return *_trace; };
```
why is this public?

abhi1287 commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -24,9 +25,11 @@ class BasicPlan
 
     AlicaTime getInterval() { return _msInterval; }
     void setInterval(int32_t msInterval) { _msInterval = AlicaTime::milliseconds(msInterval); }
+    IAlicaTrace& getTrace() { return *_trace; };
```
Why is this public? If the plan is to use this for adding child dependencies + You can return a simple std::string context  like in TraceContext.msg. rather than returning  reference to trace.

veerajsk replied:
this is for the application to set custom logs & tags inside the behaviour/plan's implementation, so we need to return the full trace object.

veerajsk replied:
or do you want the application to create a new trace for this? In that case returning the context would be sufficient

veerajsk replied:
but this will add one more nesting in the trace tree which could make trace look more visually complicated

abhi1287 replied:
ah I understand, just make this protected.

abhi1287 commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -58,6 +62,8 @@ class BasicPlan
     const Configuration* _configuration;
     AlicaTime _msInterval;
     int64_t _activeRunJobId;
+    std::unique_ptr<IAlicaTrace> _trace;
```
optional

veerajsk replied:
the trace factory returns a unique_ptr to the trace object so using optional will be unnecessary & end up consuming more space

abhi1287 commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -138,6 +138,11 @@ const AlicaClock& AlicaEngine::getAlicaClock() const
 
+const IAlicaTraceFactory* AlicaEngine::getTraceFactory() const
+{
+    return _ctx.getTraceFactory();
```
Put a comment in header that this could be null if no trace factory is set.

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -112,6 +113,9 @@ void BasicBehaviour::initJob()
     // Todo: Optimization: don't call initialiseParameters if not in context
     try {
+        if (_trace) {
+            _trace->setTag("Init", "true");
```
should be a log
_trace->setLog("Init")

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -128,6 +132,9 @@ void BasicBehaviour::runJob(void* msg)
     // TODO: get rid of msg
     try {
+        if (_trace) {
+            _trace->setTag("Run", "true");
```
can skip run since it could be called at higher frequency

veerajsk replied:
I wanted to log the first run call, not all of them. I think it could be helpful in debugging

abhi1287 replied:
We can put a one time log then. setting tag on every run isn't really efficient.

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -157,6 +164,10 @@ void BasicBehaviour::terminateJob()
     // TODO: Optimization: don't call onTermination if initiliaseParameters in not called because we were not in context
     try {
+        if (_trace) {
+            _trace->setTag("Terminate", "true");
```
log

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -185,4 +196,15 @@ bool BasicBehaviour::getParameter(const std::string& key, std::string& valueOut)
     }
 }
 
+void BasicBehaviour::startTrace()
```
Add both tags and logs in setSuccess and setFailure methods.
Ex: - 
setTag("Result", "Success" or "Fail");
setLog("Success" or "Fail");

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -23,6 +23,9 @@ void BasicPlan::doInit()
     ++_execState;
     try {
+        if (_trace) {
+            _trace->setTag("Init", "true");
```
log

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -36,6 +39,9 @@ void BasicPlan::doInit()
 {
     try {
+        if (_trace) {
+            _trace->setTag("Run", "true");
```
skip

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -83,4 +94,19 @@ void BasicPlan::stop()
+        _trace = _ae->getTraceFactory()->create(_name, parentPlan->getTrace().context());
+    }
+}
+
```
Trace success and failure. see comment for basicBehaviour

abhi1287 reviewed (commented):
we should also log when a task assignment changed or a runtime condition failed. It can happen in a separate pr though.

bjoernschroeder committed:
add comment to getTraceFactory method

bjoernschroeder committed:
use setLog, log run once, make getTrace() protected

veerajsk committed:
Merge branch 'rr-devel' into bas_tracing

veerajsk committed:
Fixes in creating traces

- Create traces in the execution context during init.
- Provide the option to enable/disable tracing

veerajsk committed:
Check if trace factory is setting before using it

veerajsk committed:
Merge branch 'rr-devel' into bas_plan_pool

veerajsk committed:
Merge branch 'bas_plan_pool' into bas_tracing

veerajsk committed:
Set basicplan name

veerajsk committed:
Merge branch 'rr-devel' into bas_tracing

veerajsk committed:
Bug fixes

- check if basic plan exists before accessing the trace context
- if no parent is found create a root trace
- remove hack to create basic plan

veerajsk committed:
Fix alica engine termination

Do the following in order to correctly terminate the engine:
1. Stop the plan base thread. This prevents any more changes to the
running plan.
2. Deactivate the running plan in plan base stop(). This ensures the
plans/behaviours are scheduled for termination in the correct order.
3. Stop all behaviours & plans that may be running. There should not
be any since (2) ensures this taken care of
4. Execute all pending non repeatable jobs in the scheduler thread.
This ensures the plan/behaviour init/terminates are actually executed.
5. Stop the scheduler thread

Add a trace log that indicates if the trace is for a behaviour or a
plan.

Note: the master plan trace is not reported because the master plan does
not have a corresponding basic plan attached to it. This is a bug &
the reason for it is because there is no ConfAbstractPlanWrapper for the
master plan due to which it is not in the plan pool. This has to be fixed
in a separate PR.

veerajsk committed:
Merge branch 'rr-devel' into bas_tracing

veerajsk committed:
Bring back scheduler null check in engine terminate

veerajsk committed:
Avoid unused variable warning

veerajsk commented:
Replaced by https://github.com/rapyuta-robotics/alica/pull/176
